core/src/main/java/ai/timefold/solver/core/impl/solver/AssignmentProcessor.java

```diff
@@ -0,0 +1,158 @@
+import ai.timefold.solver.core.impl.constructionheuristic.DefaultConstructionHeuristicPhase;
+import ai.timefold.solver.core.impl.constructionheuristic.placer.EntityPlacer;
+import ai.timefold.solver.core.impl.constructionheuristic.scope.ConstructionHeuristicPhaseScope;
+import ai.timefold.solver.core.impl.constructionheuristic.scope.ConstructionHeuristicStepScope;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.BasicVariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableDemand;
+import ai.timefold.solver.core.impl.heuristic.move.Move;
+import ai.timefold.solver.core.impl.heuristic.selector.list.LocationInList;
+import ai.timefold.solver.core.impl.heuristic.selector.move.generic.ChangeMove;
+import ai.timefold.solver.core.impl.heuristic.selector.move.generic.chained.ChainedChangeMove;
+import ai.timefold.solver.core.impl.heuristic.selector.move.generic.list.ListUnassignMove;
+import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+
+final class AssignmentProcessor<Solution_, Score_ extends Score<Score_>, Recommendation_, In_, Out_>
+        implements Function<InnerScoreDirector<Solution_, Score_>, List<Recommendation_>> {
+
+    private final DefaultSolverFactory<Solution_> solverFactory;
+    private final Function<In_, Out_> valueResultFunction;
+    private final RecommendationConstructor<Score_, Recommendation_, Out_> recommendationConstructor;
+    private final ScoreAnalysisFetchPolicy fetchPolicy;
+    private final ScoreAnalysis<Score_> originalScoreAnalysis;
+    private final In_ clonedElement;
+
+    public AssignmentProcessor(DefaultSolverFactory<Solution_> solverFactory, Function<In_, Out_> valueResultFunction,
+            RecommendationConstructor<Score_, Recommendation_, Out_> recommendationConstructor,
+            ScoreAnalysisFetchPolicy fetchPolicy, In_ clonedElement, ScoreAnalysis<Score_> originalScoreAnalysis) {
+        this.solverFactory = Objects.requireNonNull(solverFactory);
+        this.valueResultFunction = valueResultFunction;
+        this.recommendationConstructor = Objects.requireNonNull(recommendationConstructor);
+        this.fetchPolicy = Objects.requireNonNull(fetchPolicy);
+        this.originalScoreAnalysis = Objects.requireNonNull(originalScoreAnalysis);
+        this.clonedElement = clonedElement;
+    }
+
+    @Override
+    public List<Recommendation_> apply(InnerScoreDirector<Solution_, Score_> scoreDirector) {
+        // The cloned element may already be assigned.
+        // If it is, we need to unassign it before we can run the construction heuristic.
+        var supplyManager = scoreDirector.getSupplyManager();
+        var solutionDescriptor = solverFactory.getSolutionDescriptor();
+        var listVariableDescriptor = solutionDescriptor.getListVariableDescriptor();
+        if (listVariableDescriptor != null) {
```

Comment:
I'm worried that this action will not be undone. We can unassign the element, but I can't find the corresponding undo action. Could this lead to an inconsistent state?

Reply:
The action will not be undone. It doesn't need to be.
The solution at this point is already cloned - so there is no external impact.
And we do not need to undo this change for internal reasons either - the CH needs to always start from this new state.